File: experiments/mathexpr-prompt-python/mathexpr.py
"""
Math expression evaluator.

Pipeline: tokenize -> parse (recursive descent) -> evaluate AST.
"""

from __future__ import annotations
from enum import Enum, auto
from typing import List


# ---------------------------------------------------------------------------
# Tokenizer
# ---------------------------------------------------------------------------

class TokenKind(Enum):
    NUMBER = auto()
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    DOUBLESTAR = auto()
    LPAREN = auto()
    RPAREN = auto()
    EOF = auto()


class Token:
    __slots__ = ("kind", "value")

    def __init__(self, kind: TokenKind, value: str = "") -> None:
        self.kind = kind
        self.value = value

    def __repr__(self) -> str:
        return f"Token({self.kind}, {self.value!r})"


def tokenize(expr: str) -> List[Token]:
    tokens: List[Token] = []
    i = 0
    n = len(expr)

    while i < n:
        ch = expr[i]

        # whitespace
        if ch in " \t\r\n":
            i += 1
            continue

        # numbers: digits or leading dot for decimals like .5
        if ch.isdigit() or (ch == "." and i + 1 < n and expr[i + 1].isdigit()):
            start = i
            has_dot = ch == "."
            i += 1
            while i < n and (expr[i].isdigit() or (expr[i] == "." and not has_dot)):
                if expr[i] == ".":
                    has_dot = True
                i += 1
            tokens.append(Token(TokenKind.NUMBER, expr[start:i]))
            continue

        # two-char operator **
        if ch == "*" and i + 1 < n and expr[i + 1] == "*":
            tokens.append(Token(TokenKind.DOUBLESTAR, "**"))
            i += 2
            continue

        simple = {
            "+": TokenKind.PLUS,
            "-": TokenKind.MINUS,
            "*": TokenKind.STAR,
            "/": TokenKind.SLASH,
            "%": TokenKind.PERCENT,
            "(": TokenKind.LPAREN,
            ")": TokenKind.RPAREN,
        }

        if ch in simple:
            tokens.append(Token(simple[ch], ch))
            i += 1
            continue

        raise ValueError(f"Invalid character: {ch!r}")

    tokens.append(Token(TokenKind.EOF))
    return tokens


# ---------------------------------------------------------------------------
# AST nodes
# ---------------------------------------------------------------------------

class Num:
    __slots__ = ("value",)

    def __init__(self, value: float) -> None:
        self.value = value


class UnaryOp:
    __slots__ = ("op", "operand")

    def __init__(self, op: str, operand) -> None:
        self.op = op
        self.operand = operand


class BinOp:
    __slots__ = ("left", "op", "right")

    def __init__(self, left, op: str, right) -> None:
        self.left = left
        self.op = op
        self.right = right


# ---------------------------------------------------------------------------
# Parser — recursive descent
# ---------------------------------------------------------------------------

class Parser:
    def __init__(self, tokens: List[Token]) -> None:
        self.tokens = tokens
        self.pos = 0

    def peek(self) -> Token:
        return self.tokens[self.pos]

    def advance(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expect(self, kind: TokenKind) -> Token:
        tok = self.advance()
        if tok.kind != kind:
            raise ValueError(f"Expected {kind}, got {tok.kind}")
        return tok

    # entry
    def parse(self):
        node = self.expr()
        if self.peek().kind != TokenKind.EOF:
            raise ValueError("Unexpected token after expression")
        return node

    # expr -> term (('+' | '-') term)*
    def expr(self):
        node = self.term()
        while self.peek().kind in (TokenKind.PLUS, TokenKind.MINUS):
            op = self.advance().value
            right = self.term()
            node = BinOp(node, op, right)
        return node

    # term -> exponent (('*' | '/' | '%') exponent)*
    def term(self):
        node = self.exponent()
        while self.peek().kind in (TokenKind.STAR, TokenKind.SLASH, TokenKind.PERCENT):
            op = self.advance().value
            right = self.exponent()
            node = BinOp(node, op, right)
        return node

    # exponent -> unary ('**' exponent)?   (right-associative via recursion)
    def exponent(self):
        node = self.unary()
        if self.peek().kind == TokenKind.DOUBLESTAR:
            self.advance()
            right = self.exponent()  # right-recursive for right-assoc
            node = BinOp(node, "**", right)
        return node

    # unary -> '-' unary | primary
    def unary(self):
        if self.peek().kind == TokenKind.MINUS:
            self.advance()
            operand = self.unary()
            return UnaryOp("-", operand)
        return self.primary()

    # primary -> NUMBER | '(' expr ')'
    def primary(self):
        tok = self.peek()
        if tok.kind == TokenKind.NUMBER:
            self.advance()
            return Num(float(tok.value))
        if tok.kind == TokenKind.LPAREN:
            self.advance()
            node = self.expr()
            self.expect(TokenKind.RPAREN)
            return node
        raise ValueError(f"Unexpected token: {tok}")


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------

def evaluate(node) -> float:
    if isinstance(node, Num):
        return node.value
    if isinstance(node, UnaryOp):
        val = evaluate(node.operand)
        if node.op == "-":
            return -val
        raise ValueError(f"Unknown unary op: {node.op}")
    if isinstance(node, BinOp):
        left = evaluate(node.left)
        right = evaluate(node.right)
        if node.op == "+":
            return left + right
        if node.op == "-":
            return left - right
        if node.op == "*":
            return left * right
        if node.op == "/":
            if right == 0:
                raise ZeroDivisionError("Division by zero")
            return left / right
        if node.op == "%":
            if right == 0:
                raise ZeroDivisionError("Modulo by zero")
            return left % right
        if node.op == "**":
            return left ** right
        raise ValueError(f"Unknown binary op: {node.op}")
    raise ValueError(f"Unknown node type: {type(node)}")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def calc(expression: str) -> float:
    """Evaluate a math expression string and return the numeric result."""
    if not expression or not expression.strip():
        raise ValueError("Empty input")
    tokens = tokenize(expression)
    parser = Parser(tokens)
    tree = parser.parse()
    return evaluate(tree)

File: research/experiments/mathexpr-spec-go/mathexpr.go
// Package mathexpr implements a math expression parser and evaluator.
//
// It provides a pipeline of Tokenize -> Parse -> Evaluate, composed
// together via the top-level Calc function.
package mathexpr

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Token types
// ---------------------------------------------------------------------------

// TokenKind represents the type of a lexical token.
type TokenKind string

const (
	TokenNumber  TokenKind = "number"
	TokenPlus    TokenKind = "plus"
	TokenMinus   TokenKind = "minus"
	TokenStar    TokenKind = "star"
	TokenSlash   TokenKind = "slash"
	TokenPercent TokenKind = "percent"
	TokenPower   TokenKind = "power"
	TokenLParen  TokenKind = "lparen"
	TokenRParen  TokenKind = "rparen"
)

// Token is a lexical token with a kind and string value.
type Token struct {
	Kind  TokenKind
	Value string
}

// ---------------------------------------------------------------------------
// AST node types
// ---------------------------------------------------------------------------

// Node is the interface implemented by all AST nodes.
type Node interface {
	nodeTag()
}

// NumberLiteral is a leaf node holding a numeric value.
type NumberLiteral struct {
	Value float64
}

func (NumberLiteral) nodeTag() {}

// UnaryExpr is a unary operator applied to an operand (only "-" is supported).
type UnaryExpr struct {
	Op      string
	Operand Node
}

func (UnaryExpr) nodeTag() {}

// BinaryExpr is a binary operator applied to left and right operands.
type BinaryExpr struct {
	Op    string
	Left  Node
	Right Node
}

func (BinaryExpr) nodeTag() {}

// ---------------------------------------------------------------------------
// Tokenizer
// ---------------------------------------------------------------------------

// Tokenize converts a math expression string into a sequence of tokens.
func Tokenize(input string) ([]Token, error) {
	var tokens []Token
	i := 0
	runes := []rune(input)

	for i < len(runes) {
		ch := runes[i]

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		// Numbers: digits or leading dot
		if ch >= '0' && ch <= '9' || ch == '.' {
			start := i
			hasDot := false
			for i < len(runes) && (runes[i] >= '0' && runes[i] <= '9' || runes[i] == '.') {
				if runes[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("Unexpected character `.`")
					}
					hasDot = true
				}
				i++
			}
			tokens = append(tokens, Token{Kind: TokenNumber, Value: string(runes[start:i])})
			continue
		}

		switch ch {
		case '+':
			tokens = append(tokens, Token{Kind: TokenPlus, Value: "+"})
			i++
		case '-':
			tokens = append(tokens, Token{Kind: TokenMinus, Value: "-"})
			i++
		case '*':
			if i+1 < len(runes) && runes[i+1] == '*' {
				tokens = append(tokens, Token{Kind: TokenPower, Value: "**"})
				i += 2
			} else {
				tokens = append(tokens, Token{Kind: TokenStar, Value: "*"})
				i++
			}
		case '/':
			tokens = append(tokens, Token{Kind: TokenSlash, Value: "/"})
			i++
		case '%':
			tokens = append(tokens, Token{Kind: TokenPercent, Value: "%"})
			i++
		case '(':
			tokens = append(tokens, Token{Kind: TokenLParen, Value: "("})
			i++
		case ')':
			tokens = append(tokens, Token{Kind: TokenRParen, Value: ")"})
			i++
		default:
			return nil, fmt.Errorf("Unexpected character `%c` at position %d", ch, i)
		}
	}

	return tokens, nil
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

type parser struct {
	tokens []Token
	pos    int
}

func (p *parser) peek() *Token {
	if p.pos < len(p.tokens) {
		return &p.tokens[p.pos]
	}
	return nil
}

func (p *parser) advance() Token {
	tok := p.tokens[p.pos]
	p.pos++
	return tok
}

func (p *parser) expect(kind TokenKind) (Token, error) {
	tok := p.peek()
	if tok == nil {
		return Token{}, fmt.Errorf("Unexpected end of input")
	}
	if tok.Kind != kind {
		return Token{}, fmt.Errorf("Expected %s", kind)
	}
	return p.advance(), nil
}

// Parse converts a token slice into an AST.
func Parse(tokens []Token) (Node, error) {
	if len(tokens) == 0 {
		return nil, fmt.Errorf("Unexpected end of input")
	}
	p := &parser{tokens: tokens}
	node, err := p.parseAddSub()
	if err != nil {
		return nil, err
	}
	if p.peek() != nil {
		return nil, fmt.Errorf("Unexpected token after expression")
	}
	return node, nil
}

// Level 1: + and - (left-associative)
func (p *parser) parseAddSub() (Node, error) {
	left, err := p.parseMulDivMod()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || (tok.Kind != TokenPlus && tok.Kind != TokenMinus) {
			break
		}
		op := p.advance()
		right, err := p.parseMulDivMod()
		if err != nil {
			return nil, err
		}
		left = BinaryExpr{Op: op.Value, Left: left, Right: right}
	}
	return left, nil
}

// Level 2: *, /, % (left-associative)
func (p *parser) parseMulDivMod() (Node, error) {
	left, err := p.parsePower()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || (tok.Kind != TokenStar && tok.Kind != TokenSlash && tok.Kind != TokenPercent) {
			break
		}
		op := p.advance()
		right, err := p.parsePower()
		if err != nil {
			return nil, err
		}
		left = BinaryExpr{Op: op.Value, Left: left, Right: right}
	}
	return left, nil
}

// Level 3: ** (right-associative)
func (p *parser) parsePower() (Node, error) {
	base, err := p.parseUnary()
	if err != nil {
		return nil, err
	}
	tok := p.peek()
	if tok != nil && tok.Kind == TokenPower {
		p.advance()
		exp, err := p.parsePower() // right-recursive for right-associativity
		if err != nil {
			return nil, err
		}
		return BinaryExpr{Op: "**", Left: base, Right: exp}, nil
	}
	return base, nil
}

// Level 4: unary - (prefix, right-associative, can chain)
func (p *parser) parseUnary() (Node, error) {
	tok := p.peek()
	if tok != nil && tok.Kind == TokenMinus {
		p.advance()
		operand, err := p.parseUnary()
		if err != nil {
			return nil, err
		}
		return UnaryExpr{Op: "-", Operand: operand}, nil
	}
	return p.parseAtom()
}

// Level 5: atoms — numbers and parenthesized expressions
func (p *parser) parseAtom() (Node, error) {
	tok := p.peek()
	if tok == nil {
		return nil, fmt.Errorf("Unexpected end of input")
	}

	switch tok.Kind {
	case TokenNumber:
		p.advance()
		val, err := strconv.ParseFloat(tok.Value, 64)
		if err != nil {
			return nil, fmt.Errorf("invalid number: %s", tok.Value)
		}
		return NumberLiteral{Value: val}, nil

	case TokenLParen:
		p.advance()
		node, err := p.parseAddSub()
		if err != nil {
			return nil, err
		}
		if _, err := p.expect(TokenRParen); err != nil {
			return nil, fmt.Errorf("Expected rparen")
		}
		return node, nil

	default:
		return nil, fmt.Errorf("Unexpected token: %s", tok.Kind)
	}
}

// ---------------------------------------------------------------------------
// Evaluator
// ---------------------------------------------------------------------------

// Evaluate recursively evaluates an AST node and returns the numeric result.
func Evaluate(node Node) (float64, error) {
	switch n := node.(type) {
	case NumberLiteral:
		return n.Value, nil
	case UnaryExpr:
		val, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -val, nil
	case BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}
		switch n.Op {
		case "+":
			return left + right, nil
		case "-":
			return left - right, nil
		case "*":
			return left * right, nil
		case "/":
			if right == 0 {
				return 0, fmt.Errorf("Division by zero")
			}
			return left / right, nil
		case "%":
			if right == 0 {
				return 0, fmt.Errorf("Modulo by zero")
			}
			return math.Mod(left, right), nil
		case "**":
			return math.Pow(left, right), nil
		default:
			return 0, fmt.Errorf("unknown operator: %s", n.Op)
		}
	default:
		return 0, fmt.Errorf("unknown node type: %T", node)
	}
}

// ---------------------------------------------------------------------------
// Calc — top-level API
// ---------------------------------------------------------------------------

// Calc evaluates a math expression string and returns the numeric result.
// It composes Tokenize, Parse, and Evaluate into a single pipeline.
func Calc(expression string) (float64, error) {
	if strings.TrimSpace(expression) == "" {
		return 0, fmt.Errorf("Empty expression")
	}

	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}

	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}

	return Evaluate(ast)
}
